test/test-plan/eventing/content-based-routing.md

```diff
@@ -0,0 +1,143 @@
+> > (where durability means that the Broker can retry event delivery beyond the
+> > duration of receiving the event).
+
+> > For each event received by the Broker, the Broker MUST evaluate each
+> > associated Trigger exactly once (where "associated" means a Trigger with a
+> > spec.broker which references the Broker). If the Trigger has a Ready
+> > condition of true when the event is evaluated, the Broker MUST evaluate the
+> > Trigger's spec.filter and, if matched, proceed with event delivery as
+> > described below. The Broker MAY also evaluate and forward events to
+> > associated Triggers for which the Ready condition is not currently true.
+> > (One example: a Trigger which is in the process of being programmed in the
+> > Broker data plane might receive some events before the data plane
+> > programming was complete and the Trigger was updated to set the Ready
+> > condition to true.)
+
+> > If multiple Triggers match an event, one event delivery MUST be generated
+> > for each match; duplicate matches with the same destination MUST each
+> > generate separate event delivery attempts, one per Trigger match. The
+> > implementation MAY attach additional event attributes or other metadata
+> > distinguishing between these deliveries. The implementation MUST NOT modify
+> > the event data in this process.
+
+> > Reply events generated during event delivery MUST be re-enqueued by the
+> > Broker using the same routing and persistence as events delivered to the
+> > Broker's Addressable URL. Reply events re-enqueued in this manner MUST be
+> > evaluated against all Triggers associated with the Broker, including the
+> > Trigger that generated the reply. If the storage of the reply event in the
+> > Broker fails, the entire event delivery MUST be failed and the delivery to
+> > the Trigger's subscriber MUST be retried. Implementations MAY implement
+> > event-loop detection; it is RECOMMENDED that any such controls be documented
+> > to end-users. Implementations MAY avoid using HTTP to deliver event replies
+> > to the Broker's event-delivery input and instead use an internal queueing
+> > mechanism.
+
+# Testing Content Based Routing Conformance:
+
+We are going to be testing that, under normal operation (no pod failures), when
+an event is sent to a broker, each associated Trigger receives the event exactly
+once when it passes the trigger filter. We are going to test different filters.
+
+We are going to send events to the event-display service to manually check that
+events are being received.
+
+You can find the resources for running these tests inside the
+[control-plane/content-based-routing](control-plane/content-based-routing)
+directory.
+
+## [Pre] Creating a Broker
+
+```sh
+kubectl apply -f control-plane/content-based-routing/trigger.yaml
+```
+
+Expected output:
+
+```sh
+broker.eventing.knative.dev/conformance-broker created
+```
+
+## [Pre] Creating an Event Display service
```

Comment:
@lionelvillard I truly believe that we are on the right track here, but as you might have noticed we quickly run in the situation where we need a pod that basically count events, or maybe even fail at receiving events on purpose. Instead of using existing events visualizers, I was thinking to create a very specific event consumer service with fixed endpoints to accept and report back the status of the events that are delivered to it. 
I've started doing some of the data plane stuff to try to define this service more concretely.  As reference: https://github.com/salaboy/knative-conformance-experiment/blob/main/eventing/event-ack-and-retry.md 

What I would like to see here is something like: 
1) `curl` or `kn` a CloudEvent to a Broker
2)  `curl` a known endpoint (for example with the name of the test) and get the event(s) that I've received or the ones that I've failed to receive 

The main reason to aim for a simple solution here is to keep automation simple and dependency-free. I would like to rely only on the following deps: 
- https://github.com/kubernetes/client-go (not Knative Clients)
- https://github.com/cloudevents/sdk-go (to produce cloud events) 
- HTTP get and Post to interact with custom pods 

If we have a solid test plan with scenarios and test well defined then having that custom service with custom endpoints shouldn't be difficult to implement. I've started doing this here: https://github.com/salaboy/knative-conformance-experiment/tree/main/events-counter I really want to avoid parsing `kubectl logs` 

What do you think? Maybe we can gather opinions from other folks here as well.

Reply:
A more concrete example about the way that we should consume events check: 
- https://github.com/salaboy/knative-conformance-experiment/blob/main/events-counter/main.go#L23 Specific endpoint for a data plane test
- being used by this trigger: https://github.com/salaboy/knative-conformance-experiment/blob/main/eventing/data-plane/event-ack-and-retry/trigger.yaml
- That comes from this section of the test plan: https://github.com/salaboy/knative-conformance-experiment/blob/main/eventing/event-ack-and-retry.md#testing-delivery-retry-for-a-broker
- And we can curl to validate the output: https://github.com/salaboy/knative-conformance-experiment/blob/main/events-counter/main.go#L24

Reply:
sure event-display is just to get started. Have you looked at recordevents: https://github.com/knative/eventing/tree/main/test/lib/recordevents. We use this intensively for our e2e tests.